L1Trigger/TrackFindingTracklet/src/KFin.cc

```diff
@@ -0,0 +1,424 @@
+#include "L1Trigger/TrackFindingTracklet/interface/KFin.h"
+
+#include <vector>
+#include <numeric>
+#include <algorithm>
+
+using namespace std;
+using namespace edm;
+using namespace tt;
+using namespace trackerTFP;
+
+namespace trklet {
+
+  KFin::KFin(const ParameterSet& iConfig,
+             const Setup* setup,
+             const DataFormats* dataFormats,
+             const LayerEncoding* layerEncoding,
+             const ChannelAssignment* channelAssignment,
+             const Settings* settings,
+             int region)
+      : enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+        setup_(setup),
+        dataFormats_(dataFormats),
+        layerEncoding_(layerEncoding),
+        channelAssignment_(channelAssignment),
+        settings_(settings),
+        region_(region),
+        input_(channelAssignment_->numChannelsTrack()) {
+    // unified tracklet bases
+    baseUinv2R_ = .5 * settings_->kphi1() / settings_->kr() * pow(2, settings_->rinv_shift());
+    baseUphiT_ = settings_->kphi1() * pow(2, settings_->phi0_shift());
+    baseUcot_ = settings_->kz() / settings_->kr() * pow(2, settings_->t_shift());
+    baseUzT_ = settings_->kz() * pow(2, settings_->z0_shift());
+    baseUr_ = settings_->kr();
+    baseUphi_ = settings_->kphi1();
+    baseUz_ = settings_->kz();
+    // low precise bases KF is using
+    baseLinv2R_ = dataFormats->base(Variable::inv2R, Process::kfin);
+    baseLphiT_ = dataFormats->base(Variable::phiT, Process::kfin);
+    baseLcot_ = dataFormats->base(Variable::cot, Process::kfin);
+    baseLzT_ = dataFormats->base(Variable::zT, Process::kfin);
+    baseLr_ = dataFormats->base(Variable::r, Process::kfin);
+    baseLphi_ = dataFormats->base(Variable::phi, Process::kfin);
+    baseLz_ = dataFormats->base(Variable::z, Process::kfin);
+    // higher precision bases used to transform unified to low bases
+    baseHinv2R_ = baseLinv2R_ * pow(2, floor(log2(baseUinv2R_ / baseLinv2R_)));
+    baseHphiT_ = baseLphiT_ * pow(2, floor(log2(baseUphiT_ / baseLphiT_)));
+    baseHcot_ = baseLcot_ * pow(2, floor(log2(baseUcot_ / baseLcot_)));
+    baseHzT_ = baseLzT_ * pow(2, floor(log2(baseUzT_ / baseLzT_)));
+    baseHr_ = baseLr_ * pow(2, floor(log2(baseUr_ / baseLr_)));
+    baseHphi_ = baseLphi_ * pow(2, floor(log2(baseUphi_ / baseLphi_)));
+    baseHz_ = baseLz_ * pow(2, floor(log2(baseUz_ / baseLz_)));
+    // calculate base used for inverted cot(theta)
+    const int baseShiftInvCot = ceil(log2(setup_->outerRadius() / setup_->hybridRangeR())) - setup_->widthDSPbu();
+    baseInvCot_ = pow(2, baseShiftInvCot);
+  }
+
+  // read in and organize input tracks and stubs
+  void KFin::consume(const StreamsTrack& streamsTrack, const StreamsStub& streamsStub) {
+    static constexpr int widthPhi = 12;
```

Comment:
should these all be set in a config file?

Reply:
1) You've now added "WidthPhi = ... 12 ..." etc. to TrackTrigger/python/ProducerSetup_cfi.py to specify the "number of bits used at TrackBuilder output". This duplicates the configuration info in Settings.h (int phiresidbits_{12} etc.). Hence I assume that anyone changing this parameter must remember to change it in two places?
2) Both your cfg params and those in Settings.h set the number of phi bits to 12, which differs from the number for the TrackBuilder output specified in
 https://twiki.cern.ch/twiki/bin/view/CMS/HybridDataFormat#TrackWord_StubWord_written_by_Tr , in particular because the latter is different in barrel & endcap. Any idea why?

Reply:
I found now all numbers in settings and use those, the twiki page says 12 bits in all cases